src/term/net_to_term.rs

```diff
@@ -327,6 +352,51 @@ impl<'a> Reader<'a> {
   }
 }
 
+/// Recover string constructors when it is not possible to correctly readback a string
+fn recover_constructors(term: &mut Term, cons_term: Term) {
+  match term {
+    Term::Str { val } => {
+      let snil = Term::Var { nam: Name::new(SNIL) };
+      let last_term = string_cons(cons_term, snil);
+      *term = string_app(val, last_term)
+    }
+    app @ Term::App { .. } => insert_string_cons(app, cons_term),
+    _ => unreachable!(),
+  }
+}
+
+// TODO: `λa (SCons a (SCons 'b' (SCons 'c' SNil)))` could be read back as `λa (SCons a "bc")`
```

Comment:
Why not simply try twice?

In `resugar_adts`:
```
match term {
  ...
  Term::Lam {...} if tag == Tag::string => {
    // The second one recreates the constructor and calls resugar_adts recursively on the ctr fields. Not necessarily a separate function, just for the sake of keeping this comment short
    resugar_str(term).or_else(|| rebuild_constructor(term)) 
  }
  ...
}
```
This would allow you to still use the old very simple resugar_str function and have the readback of substrings in a malformed string like mentioned in this todo.

Reply:
Also, I feel like we're polluting this file too much and it's getting a bit out of hand (not due to this PR specifically but it's growing and getting messier over time).

Reply:
Another alternative would be to first resugar all constructors ignoring the existence of lists and strings, and then right after try to construct the list and string sugars from the constructors. That would also make the code a bit cleaner i think
```
// in resugar_adts
match term {
  ...
  Term::Lam {tag, ...} => {
    if book.adt.contains(tag) {
      // Tries to recover this constructor and calls resugar_adt recursively
      *term = rebuild_constructors(body);
      match tag {
        // Since this function is recursive, if the constructor fields were lists or strings, they will already have been resugared, making these 2 resugar_functions simpler.
        Tag::string => resugar_string(term)
        Tag::list => resugar_list(term)
      }
    }
  }
  ...
}
```

Reply:
For now I'll accept it, but I'll open a separate story for us to refactor it